### PR TITLE
[JENKINS-21311] Make the queue/cancelItem API an API.

### PR DESCRIPTION
**Pull request by PierreBtz**
This change:
* documents the API
* adds meaningful return codes to help API caller.

See [JENKINS-21311](https://issues.jenkins-ci.org/browse/JENKINS-21311).

### Proposed changelog entries

* Entry 1: Internal: Make the `queue/cancelItem` return meaningful codes instead of a 404.
* Entry 2: Internal: Document the `queue/cancelItem` endpoint.

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- ~[ ] For dependency updates: links to external changelogs and, if possible, full diffs~

### CODE REVIEW

File: core/src/main/java/hudson/model/Queue.java [hunk 1]
```diff
@@ -760,17 +762,21 @@ public boolean cancel(Item item) {
...
-                cancel(item);
+                if(cancel(item)) {
+                    return HttpResponses.status(HttpServletResponse.SC_NO_CONTENT);
+                }
+                return HttpResponses.error(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, "Could not cancel run for id " + id);
             }
+            return HttpResponses.error(422, "Item for id (" + id + ") is not cancellable");
         } // else too late, ignore (JENKINS-14813)
-        return HttpResponses.forwardToPreviousPage();
+        return HttpResponses.error(422, "Provided id (" + id + ") does not exist");
```
Comment (timja) [inline core/src/main/java/hudson/model/Queue.java]: 404 instead?

Comment (res0nance) [reply core/src/main/java/hudson/model/Queue.java]: Seconded it should be 404. if the id is provided the request is not malformed

Comment (PierreBtz) [reply core/src/main/java/hudson/model/Queue.java]: Fixed!

File: core/src/main/java/hudson/model/Queue.java [hunk 2]
```diff
@@ -2303,11 +2309,11 @@ public String getSearchUrl() {
         /** @deprecated Use {@link #doCancelItem} instead. */
         @Deprecated
         @RequirePOST
-        public HttpResponse doCancelQueue() throws IOException, ServletException {
+        public HttpResponse doCancelQueue() {
             if(hasCancelPermission()){
                 Jenkins.get().getQueue().cancel(this);
             }
-            return HttpResponses.forwardToPreviousPage();
+            return HttpResponses.status(HttpServletResponse.SC_NO_CONTENT);
```
Comment (varyvol) [inline core/src/main/java/hudson/model/Queue.java]: Is this used anywhere to be called from the UI or something similar? I'm afraid that stop calling forwardToPrevious() changes the behaviour for the cases where this is not expected to behave 100% like an endpoint.

**Review by varyvol — COMMENTED** [review]
If you could check that cancelling items from the UI is still working properly, I think that is important. Also there is an old test failing because it needs to be updated with the new response codes.

File: core/src/main/java/hudson/model/Queue.java (hunk 2 above)
Comment (PierreBtz) [reply core/src/main/java/hudson/model/Queue.java]: It's called in two places in the UI and each places we don't dismiss the redirection. I did the test  canceling a job from the UI.
I just noticed I missed an integration test that I must admit I didn't run locally, I'll fix that :)

Comment (PierreBtz) [reply core/src/main/java/hudson/model/Queue.java]: Fixed

**Review by oleg-nenashev — APPROVED** [review]
I think it is a net improvement for this API endpoint, thanks!

**Review by timja — APPROVED** [review]
Looks fine

**Comment by oleg-nenashev** [comment]
I plan to merge it tomorrow if no negative feedback

File: core/src/main/resources/hudson/model/queue/_api.jelly [hunk 3]
```diff
@@ -0,0 +1,17 @@
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form" xmlns:i="jelly:fmt">
+
+  <h2>Canceling the build of an item in the Queue</h2>
+  <p>
+    To programmatically cancel the build of an item in the Queue, post to <a href="../cancelItem">this URL</a>.
+    The id of the item to cancel is expected.
```
Comment (daniel-beck) [inline core/src/main/resources/hudson/model/queue/_api.jelly]: Expected where?

Comment (PierreBtz) [reply core/src/main/resources/hudson/model/queue/_api.jelly]: Would rewording as:

"The id of the item to cancel is expected to be passed as a query parameter named id" answer your question?

Comment (daniel-beck) [reply core/src/main/resources/hudson/model/queue/_api.jelly]: @PierreBtz That would work 👍
